dynamiqs/time_array.py

```diff
@@ -390,48 +406,46 @@ def __add__(self, other: ArrayLike | TimeArray) -> TimeArray:
 
 
 class ModulatedTimeArray(TimeArray):
-    f: callable[[float, ...], Array]  # (...)
+    f: callable[[float], Array]  # (...)
     array: Array  # (n, n)
-    args: tuple[PyTree]
 
     @property
     def dtype(self) -> np.dtype:
         return self.array.dtype
 
     @property
     def shape(self) -> tuple[int, ...]:
-        f_shape = jax.eval_shape(self._call_f, 0.0).shape
+        f_shape = jax.eval_shape(self.f, 0.0).shape
         return (*f_shape, *self.array.shape)
 
     @property
     def mT(self) -> TimeArray:
-        return ModulatedTimeArray(self.f, self.array.mT, self.args)
-
-    def _call_f(self, t: float) -> Array:
-        args = expand_as_broadcastable(self.args)
-        return self.f(t, *args)
-
-    def __call__(self, t: float) -> Array:
-        values = self._call_f(t)
-        return values.reshape(*values.shape, 1, 1) * self.array
+        return ModulatedTimeArray(self.f, self.array.mT)
 
     def reshape(self, *new_shape: int) -> TimeArray:
-        f_shape = jax.eval_shape(self._call_f, 0.0).shape
-        new_f_shape, new_array_shape = _split_shape(
-            new_shape, f_shape, self.array.shape
-        )
-        f = Partial(lambda t, *args: self.f(t, *args).reshape(*new_f_shape))
-        return ModulatedTimeArray(f, self.array.reshape(*new_array_shape), self.args)
+        f = jtu.Partial(lambda t: self.f(t).reshape(*new_shape[:-2]))
```

Comment:
f is no longer a BatchedCallable from here onward. I'm guessing this will cause problems when vmapping later on. I think we should thus reshape the function inside the BatchedCallable?

Reply:
I don't think it'll be a problem because everything is a pytree here, so the vmapping will propagate. But it's worth checking

Reply:
There's a lambda in the middle so I'm not 100% confident, but maybe!

Reply:
To be clear, the BatchedCallable is still here, only wrapped in a Partial